src/lib/homestore_backend/hs_blob_manager.cpp

```diff
@@ -1,17 +1,271 @@
 #include "hs_homeobject.hpp"
+#include "replication_message.hpp"
+#include "replication_state_machine.hpp"
+#include "lib/homeobject_impl.hpp"
+
+SISL_LOGGING_DECL(blobmgr)
 
 namespace homeobject {
 
-BlobManager::Result< blob_id_t > HSHomeObject::_put_blob(ShardInfo const&, Blob&&) {
-    return folly::makeUnexpected(BlobError::UNKNOWN);
+BlobManager::AsyncResult< blob_id_t > HSHomeObject::_put_blob(ShardInfo const& shard, Blob&& blob) {
+    auto& pg_id = shard.placement_group;
+    shared< homestore::ReplDev > repl_dev;
+    {
+        std::shared_lock lock_guard(_pg_lock);
+        auto iter = _pg_map.find(pg_id);
+        if (iter == _pg_map.end()) {
+            LOGERROR("failed to put blob with non-exist pg [{}]", pg_id);
+            return folly::makeUnexpected(BlobError::UNKNOWN_PG);
+        }
+        repl_dev = static_cast< HS_PG* >(iter->second.get())->repl_dev_;
+    }
+
+    RELEASE_ASSERT(repl_dev != nullptr, "Repl dev instance null");
+
+    const uint32_t needed_size = sizeof(ReplicationMessageHeader);
+    auto req = repl_result_ctx< BlobManager::Result< BlobInfo > >::make(needed_size, 512);
+
+    uint8_t* raw_ptr = req->hdr_buf_.bytes;
+    ReplicationMessageHeader* header = new (raw_ptr) ReplicationMessageHeader();
+    header->msg_type = ReplicationMessageType::PUT_BLOB_MSG;
+    header->payload_size = 0;
+    header->payload_crc = 0;
+    header->shard_id = shard.id;
+    header->pg_id = pg_id;
+    header->header_crc = header->calculate_crc();
+
+    BlobHeader blob_header{};
+    blob_header.shard_id = shard.id;
+    blob_header.total_size = blob.body.size;
+    blob_header.hash_algorithm = BlobHeader::HashAlgorithm::CRC32;
+    blob_header.compute_blob_hash(blob.body, false /* verify */);
+
+    sisl::sg_list sgs;
+    sgs.size = 0;
+    auto block_size = repl_dev->get_blk_size();
+    auto size = sisl::round_up(sizeof(blob_header), block_size);
+    auto buf_header = iomanager.iobuf_alloc(block_size, size);
+    std::memcpy(buf_header, &blob_header, sizeof(blob_header));
+
+    // Create blob header.
+    sgs.iovs.emplace_back(iovec{.iov_base = buf_header, .iov_len = size});
+    sgs.size += size;
+
+    // Append blob bytes.
+    sgs.iovs.emplace_back(iovec{.iov_base = blob.body.bytes, .iov_len = blob.body.size});
+    sgs.size += blob.body.size;
```

Comment:
sgs.size is needed to be HS datasvc block_size aligned, otherwise the alloc blk id will assert failed.
but there is no guarantee that blob.body.size is aligned with HS blk_size, so if I tried to write a blob with size = 10 in the test case,  repl_dev->async_alloc_write() will crash.

Reply:
Had a discussion on this with hari, yaming, zhichang. Plan is to use have SM align with block_size else we will have to do lot of alloc and copies of data which affects performance and cpu.

Reply:
We dont want to copy the data, so we can reuse the memory of body and user_key for iovec.

Reply:
How SM knows the block size of HS to do the align ?  How can SM do that w/o copy?

Reply:
If gateways divides blobs of size 16/32MB, then its aligned by 1k block size, then only the last blob will be unaligned. HO can align and do memcpy. But if there are lot of small blobs unaligned, lot of copies will happen.

Reply:
GW for sure will divided blobs by a few MB (lets call it TARGET_BLOB_SIZE). 
However there will be a lot of small objects  whose size < `TARGET_BLOB_SIZE`, then they hit SM as is.

So SM/HO need to do alignment check and memcpy once if the blob is not align,  once and only once.  So I think better HO do that instead of SM.

Reply:
Yes we need to align the sgs.size to be 1k block size. So we are adding padding now after the metadata. Each iovec also needs to have address and size aligned 512 bytes due to DIRECT IO.